### PR TITLE
feat(hub-common): adds downloadFormats property to configure formats …

### PR DESCRIPTION
**Pull request by tannerjt**
…and order

affects: @esri/hub-common

ISSUES CLOSED: 8620

1. Description: Adds the ability to configure download formats visibility and order for an entity.

1. Instructions for testing:

1. Closes Issues: #8620

1. [x] Updated meaningful TSDoc to methods including Parameters and Returns, see [Documentation Guide](https://esri.github.io/hub-components/storybook/?path=/story/guides-documentation--page)

1. [x] used semantic commit messages
  
1. [x] PR title follows semantic commit format (**CRITICAL** if the title is not in a semantic format, the release automation will not run!)

1. [x] updated `peerDependencies` as needed. **CRITICAL** our automated release system can **not** be counted on to update `peerDependencies` so we _must_ do it manually in our PRs when needed. See the [updating peerDependencies](/RELEASE.md#Updating-peerDependencies) section of the release instructions for more details.

### CODE REVIEW

File: packages/common/src/core/types/IHubEditableContent.ts [hunk 1]
```diff
@@ -58,6 +58,55 @@ export interface IHubEditableContent
  */
 export type IExtendedProps = IContentExtendedProps | IServiceExtendedProps;
 
+/**
+ * Represents the download process that the configuration was created for
+ * TODO: Should we change this to DownloadFlowType?
```
Comment (abp6318) [inline packages/common/src/core/types/IHubEditableContent.ts:63]: +1 for `DownloadFlowType`

File: packages/common/src/downloads/_internal/getDownloadConfigurationDisplayFormats.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  // the criteria be met.
+  if (!flowType && ["Feature Service", "Map Service"].includes(entity.type)) {
+    const pagingFormats: IDownloadFormatConfiguration[] =
+      getPagingJobFormats().map((f) => {
+        return {
+          key: f.format,
+          hidden: false,
+        };
+      });
+    // TODO: Should we just show paging formats as a preview or include additional resources?
```
Comment (abp6318) [inline packages/common/src/downloads/_internal/getDownloadConfigurationDisplayFormats.ts:40]: highlighting a TODO -- also unsure -- might be a product question

File: packages/common/src/downloads/_internal/getDownloadConfigurationDisplayFormats.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  return formats.map((f) =>
+    isAdditionalResourceConfiguration(f)
+      ? toAdditionalResourceConfigurationDisplay(f, additionalResources)
+      : toDownloadFormatConfigurationDisplay(f)
+  );
+}
+
+// Returns true if the configuration is for an additional resource
+// TODO: Export this function as an internal helper
```
Comment (abp6318) [inline packages/common/src/downloads/_internal/getDownloadConfigurationDisplayFormats.ts:72]: highlighting a TODO

File: packages/common/src/downloads/getDownloadConfiguration.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ * If no configuration exists, a default configuration is returned based on the entity's current download flow.
+ * If a configuration exists but is no longer valid, the default configuration will also be returned.
+ *
+ * @param entity entity to get download configuration for
+ * @returns the current download configuration for the entity
+ */
+export function getDownloadConfiguration(
+  entity: IHubEditableContent
+): IEntityDownloadConfiguration {
+  // TODO: account for enterprise environments
```
Comment (abp6318) [inline packages/common/src/downloads/getDownloadConfiguration.ts:27]: highlighting TODO

File: packages/common/src/downloads/getDownloadConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    },
+    paging: () => {
+      serverFormats = getPagingJobFormats();
+    },
+    exportImage: () => {
+      serverFormats = getExportImageFormats(entity);
+    },
+  };
+
+  actionsByFlow[downloadFlow] && actionsByFlow[downloadFlow]();
```
Comment (abp6318) [inline packages/common/src/downloads/getDownloadConfiguration.ts:59]: I would opt for the helper function like you said. It's a complex part of an already meaty function.

File: packages/common/src/downloads/getDownloadFormats.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ * @param options options to refine / filter the results of the fetchDownloadFormats operation
+ * @returns The available download formats and additional resources
+ */
+export function getDownloadFormats(
+  options: IFetchDownloadFormatsOptions
+): IDownloadFormat[] {
+  const { entity, context } = options;
+  // get the base formats for the item
+  let baseFormats: IDynamicDownloadFormat[] = [];
+  // TODO: use typescript to enforce a branch for each flow type
```
Comment (abp6318) [inline packages/common/src/downloads/getDownloadFormats.ts:28]: highlighting a TODO

File: packages/common/src/downloads/getDownloadFormats.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    getProp(entity, "extendedProps.additionalResources") || [];
+  const additionalFormats = additionalResources.map(toStaticFormat);
+
+  // Respect the order and visibility of the formats as configured for the entity
+  const downloadConfiguration = getDownloadConfiguration(entity);
+  return downloadConfiguration.formats.reduce((acc, format) => {
+    if (!format.hidden) {
+      let includedFormat;
+      // TODO: use utility function to determine if the format is an additional resource
+      const isAdditionalResource = format.key.startsWith(
```
Comment (abp6318) [inline packages/common/src/downloads/getDownloadFormats.ts:46]: Don't we have a helper function for this in this PR?

File: packages/common/src/downloads/getDownloadFormats.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        "additionalResource::"
+      );
+      if (isAdditionalResource) {
+        const additionalResourceIndex = parseInt(format.key.split("::")[1], 10);
```
Comment (abp6318) [inline packages/common/src/downloads/getDownloadFormats.ts:50]: ditto here, I think I saw a helper function for this

**Review by abp6318 — COMMENTED** [review]
Looks good so far!

File: packages/common/src/downloads/getDownloadFormats.ts (hunk 7 above)
Comment (tannerjt) [reply packages/common/src/downloads/getDownloadFormats.ts]: moved to private internal helper for reuse.

File: packages/common/src/downloads/getDownloadFormats.ts (hunk 8 above)
Comment (tannerjt) [reply packages/common/src/downloads/getDownloadFormats.ts]: moved to private internal helper for reuse.

**Review by sonofflynn89 — APPROVED** [review]
Fantastic work Josh!

**Pull request merged**